include/itertools.hpp

```diff
@@ -13,6 +13,42 @@ static inline void report() {
 	puts(__PRETTY_FUNCTION__);
 }
 
+template<typename T>
+class EnumSeqIterator {
+	T _value;
+
+public:
+	explicit EnumSeqIterator(T value) : _value(value) {}
+
+	EnumSeqIterator &operator++() {
+		_value = (T)(_value + 1);
+		return *this;
+	}
+
+	auto operator*() const { return _value; }
+
+	friend auto operator==(EnumSeqIterator const &lhs, EnumSeqIterator const &rhs) {
+		return lhs._value == rhs._value;
+	}
+
+	friend auto operator!=(EnumSeqIterator const &lhs, EnumSeqIterator const &rhs) {
+		return lhs._value != rhs._value;
+	}
```

Comment:
Since we're using C++20, maybe consider using the spaceship `<=>` for this?

Reply:
When I replace `==` and `!=` with one `operator<=>`, I get a bunch of build errors.

Reply:
Oh, it has to be `= default`. And can't do that for `Zip` too because it's not just trivially doing `lhs.field <=> rhs.field`.

Reply:
...Hmm. That change worked on my machine (g++ 11.4.0) but broke CI. Oh well.

Reply:
Broke how? The check mark below is green.

Reply:
> Oh, it has to be `= default`. And can't do that for `Zip` too because it's not just trivially doing `lhs.field <=> rhs.field`.

As far as I can tell, no. [Here's an example from cppreference](https://en.cppreference.com/w/cpp/language/default_comparisons#Weak_ordering).

Reply:
> Broke how? The check mark below is green.

I undid the spaceship commit.

Reply:
Weird, I thought there'd be a record of that commit being pushed here regardless.

Reply:
I mean, feel free to commit in this pr yourself with spaceship, maybe I missed something about it.

Reply:
I mean, feel free to commit in this pr yourself with spaceship, maybe I missed something about it.

Reply:
Nonono, I meant keeping `operator==`, and only removing `operator!=`!

Reply:
Even when all I did was remove `operator!=`, CI failed with lots of "`error: no match for ‘operator!=’`".

Reply:
I'd like to just merge this as-is, and then you can see if omitting `!=` or using `<=>` instead or whatever other C++20isms will improve the codebase.

Reply:
Yeah, sure. Incremental improvements.